src/compiler/checker.ts

```diff
@@ -9401,7 +9401,30 @@ namespace ts {
                     error(node, Diagnostics.Value_of_type_0_is_not_callable_Did_you_mean_to_include_new, typeToString(funcType));
                 }
                 else {
-                    error(node, Diagnostics.Cannot_invoke_an_expression_whose_type_lacks_a_call_signature);
+                    if (apparentType.flags & TypeFlags.Union) {
+                        const unionApparentType = <UnionType>apparentType;
+                        let unionCallSignatures: Signature[] = [];
+                        let hasError = false;
+                        forEach(unionApparentType.types, type => {
+                            const eachCallSignature = getSignaturesOfType(type, SignatureKind.Call);

```

Comment:
Extract the logic here out to a function called `typeHasCallSignatures` and place it next to `typeHasConstructSignatures`. Don't base the new function on `typeHasConstructSignatures`